File: screenshot/Views/TappableTextView.m
//
//  TappableTextView.m
//  screenshot
//
//  Created by Corey Werner on 8/15/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "TappableTextView.h"
#import "screenshot-Swift.h"

@interface TappableTextView ()

@property (nonatomic, strong) NSArray<NSNumber *> *tappableIndexes;

@end

@implementation TappableTextView
@dynamic delegate;

- (instancetype)initWithFrame:(CGRect)frame textContainer:(NSTextContainer *)textContainer {
    self = [super initWithFrame:frame textContainer:textContainer];
    if (self) {
        UITapGestureRecognizer *tappableTextGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tappableTextGestureRecognizerAction:)];
        [self addGestureRecognizer:tappableTextGestureRecognizer];
    }
    return self;
}

- (void)applyTappableText:(NSArray<NSDictionary<NSString *, NSNumber *> *> *)texts {
    [self applyTappableText:texts withAttributes:nil];
}

- (void)applyTappableText:(NSArray<NSDictionary<NSString *, NSNumber *> *> *)texts withAttributes:(NSDictionary<NSString *, id> *)attributes {
    NSMutableArray *tappableIndexes = [NSMutableArray array];
    NSMutableAttributedString *attributedString = [[NSMutableAttributedString alloc] init];
    
    for (NSUInteger i = 0; i < texts.count; i++) {
        NSDictionary *dictionary = texts[i];
        NSString *text = [[dictionary allKeys] firstObject];
        BOOL isTappable = [[[dictionary allValues] firstObject] boolValue];
        NSDictionary *fragmentAttributes;
        
        if (isTappable) {
            [tappableIndexes addObject:@(i)];
            fragmentAttributes = @{[NSString stringWithFormat:@"%lu", (unsigned long)i]: @(isTappable),
                                   NSUnderlineStyleAttributeName: @(NSUnderlineStyleSingle),
                                   NSUnderlineColorAttributeName: [UIColor gray7]
                                   };
        }
        
        NSAttributedString *fragmentAttributedString = [[NSAttributedString alloc] initWithString:text attributes:fragmentAttributes];
        [attributedString appendAttributedString:fragmentAttributedString];
    }
    
    if (attributes) {
        [attributedString addAttributes:attributes range:NSMakeRange(0, attributedString.length)];
    }
    
    self.tappableIndexes = tappableIndexes;
    self.attributedText = attributedString;
}

- (void)tappableTextGestureRecognizerAction:(UITapGestureRecognizer *)tapGestureRecognizer {
    if ([self.delegate respondsToSelector:@selector(tappableTextView:tappedTextAtIndex:)]) {
        CGPoint location = [tapGestureRecognizer locationInView:self];
        location.x -= self.textContainerInset.left;
        location.y -= self.textContainerInset.top;
        
        NSUInteger characterIndex = [self.layoutManager characterIndexForPoint:location inTextContainer:self.textContainer fractionOfDistanceBetweenInsertionPoints:NULL];
        
        if (self.textStorage.length > characterIndex) {
            for (NSNumber *tappableNumber in self.tappableIndexes) {
                NSString *tappableIndexString = [NSString stringWithFormat:@"%@", tappableNumber];
                
                if ([self.attributedText attribute:tappableIndexString atIndex:characterIndex effectiveRange:NULL]) {
                    [self.delegate tappableTextView:self tappedTextAtIndex:[tappableNumber unsignedIntegerValue]];
                    break;
                }
            }
        }
    }
}

@end

File: screenshot/View/Generic/TextView/TappableTextView.swift
//
//  TappableTextView.swift
//  screenshot
//
//  Created by Corey Werner on 12/19/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

import Foundation

class TappableTextView : UITextView, TappableTextProtocol {
    weak var tappableTextDelegate: TappableTextDelegate?
    private var tappableIndexes: [UInt]?
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        
        addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(tapGestureRecognizerAction(_:))))
    }
    
    @objc private func tapGestureRecognizerAction(_ tapGesture: UITapGestureRecognizer) {
        guard let tappableIndexes = tappableIndexes, let tappableTextDelegate = tappableTextDelegate else {
            return
        }
        
        var location = tapGesture.location(in: self)
        location.x -= textContainerInset.left
        location.y -= textContainerInset.top
        
        let characterIndex = layoutManager.characterIndex(for: location, in: textContainer, fractionOfDistanceBetweenInsertionPoints: nil)
        
        if textStorage.length > characterIndex {
            for tappableIndex in tappableIndexes {
                if attributedText.attribute("\(tappableIndex)", at: characterIndex, effectiveRange: nil) != nil {
                    tappableTextDelegate.tappableText(view: self, tappedTextAt: tappableIndex)
                    break
                }
            }
        }
    }
    
    func applyTappableText(_ texts: [[String : Bool]], with attributes: [String : AnyObject]? = nil) {
        var tappableIndexes: [UInt] = []
        let attributedString = NSMutableAttributedString()
        
        texts.enumerated().forEach { (index: Int, dictionary: [String : Bool]) in
            if let text = dictionary.keys.first, let isTappable = dictionary.values.first {
                var fragmentAttributes: [String : Any]?
                
                if isTappable {
                    tappableIndexes.append(UInt(index))
                    fragmentAttributes = [
                        "\(index)": isTappable,
                        NSUnderlineStyleAttributeName: NSUnderlineStyle.styleSingle.rawValue,
                        NSUnderlineColorAttributeName: UIColor.gray7
                    ]
                }
                
                let fragmentAttributedString = NSAttributedString(string: text, attributes: fragmentAttributes)
                attributedString.append(fragmentAttributedString)
            }
        }
        
        if let attributes = attributes {
            attributedString.addAttributes(attributes, range: NSRange(location: 0, length: attributedString.length))
        }
        
        self.tappableIndexes = tappableIndexes
        attributedText = attributedString
    }
}
